### DISCUSSION TITLE
How To use Coravel With WPF Desktop App

### DISCUSSION THREAD

**Discussion by wps-preston** [jamesmh/coravel#335] [Q&A]
I apologize early if this is a foolish or basic question i am trying to use Coravel within a C# WPF NET6 desktop application I am having problem trying to determine where you put the initialization code for Coravel such as below:

Example From Website
In your .NET Core app's Startup.cs file, inside the ConfigureServices() method, add the following:

`services.AddScheduler()`

Then in the Configure() method, you can use the scheduler:

`var provider = app.ApplicationServices;
provider.UseScheduler(scheduler =>
{
    scheduler.Schedule(
        () => Console.WriteLine("Every minute during the week.")
    )
    .EveryMinute()
    .Weekday();
});`

My goal is to be able to start the scheduled tasks from several buttons withing the WPF app, I like the idea of using DI along with the Invocables but i am struggling with where is configure and insert the code above. i have been looking in the App.xaml.cs portion and overriding the OnStartup method but not sure that is the proper location.

Any help/Examples would be awesome.
Thank You

**Comment by jamesmh** [comment]
See #334  - it might be that WPF doesn't support using HostedServices, which Coravel uses under the covers :(

**Comment by holymosh** [comment]
For using Coravel with WPF app you can create host and run this host with your app
```
public class TestInvocable : IInvocable
{
    public Task Invoke()
    {
        Debug.WriteLine("TestInvocable");
        return Task.CompletedTask;
    }
}

```
But following this way your App class will contains next code. App class doesn't have xaml 
```
public class App : Application
{
    private readonly MainWindow _mainWindow;
 
    public App(MainWindow mainWindow)
    {
        this._mainWindow = mainWindow;
    }
    protected override void OnStartup(StartupEventArgs e)
    {
        _mainWindow.Show();
        base.OnStartup(e);
    }
}
```

Finally, Program class with host and app creation:
```
public class Program
{
    [STAThread]
    public static void Main(string[] args)
    {
        var host = Host.CreateDefaultBuilder()
            .ConfigureServices(services =>
            {
                services.AddSingleton<App>();
                services.AddSingleton<MainWindow>();
                services.AddSingleton<TestInvocable>();
                services.AddScheduler();
            })
            .Build();
        Console.WriteLine("Application started");
        host.Services.UseScheduler(s => s.Schedule<TestInvocable>().EverySeconds(1));
        var app = host.Services.GetService<App>();
        app.Startup += (sender, eventArgs) => host.RunAsync();
        app.Exit += (sender, eventArgs) =>  host.StopAsync();
        app.SessionEnding += (sender, eventArgs) =>  host.StopAsync();
        app.Run();
    }
}

```
![image](https://github.com/user-attachments/assets/397593cc-56eb-4247-9838-0992846ef6db)
